genie-core/src/main/java/com/netflix/genie/core/util/ExponentialBackOffTrigger.java

```diff
@@ -0,0 +1,141 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.core.util;
+
+import org.springframework.scheduling.Trigger;
+import org.springframework.scheduling.TriggerContext;
+
+import javax.annotation.concurrent.ThreadSafe;
+import java.util.Date;
+
+/**
+ * Trigger implementation whose scheduling delay grows exponentially based on a given factor.
+ * The delay is bound between some minimum and maximum.
+ * The delay can be relative to the previous task scheduling/execution/completion.
+ *
+ * @author mprimi
+ * @since 3.3.9
+ */
+@ThreadSafe
+public class ExponentialBackOffTrigger implements Trigger {
+    private final DelayType delayType;
+    private final long minDelay;
+    private final Long maxDelay;
+    private final float factor;
+    private long currentDelay;
+
+    /**
+     * Constructor.
+     * @param delayType type of delay
+     * @param minDelay minimum delay in milliseconds
+     * @param maxDelay maximum delay in milliseconds
+     * @param factor multiplier factor to grow the delay
+     */
+    public ExponentialBackOffTrigger(
+        final DelayType delayType,
+        final long minDelay,
+        final Long maxDelay,
+        final float factor
+    ) {
+        this.delayType = delayType;
+        this.minDelay = minDelay;
+        this.maxDelay = maxDelay;
+        this.factor = factor;
+
+        if (minDelay < 1) {
+            throw new IllegalArgumentException("Minimum delay must be a positive number");
```

Comment:
I think if you have these kind of preconditions you should capture them in the javadoc for the constructor otherwise you're just throwing unexpected exceptions.